src/v/jl/jls/sparse_index.cc

```diff
@@ -0,0 +1,78 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "jl/jls/sparse_index.h"
+
+#include "base/likely.h"
+#include "base/vassert.h"
+#include "jl/jls/jls.h"
+#include "model/fundamental.h"
+
+#include <absl/types/optional.h>
+
+#include <optional>
+
+namespace jl::jls {
+
+std::optional<chunk_loc>
+sparse_index::find_chunk(const model::ntp& ntp, model::offset offset) const {
+    vassert(!_data.entries.empty(), "jls chunk index is empty");
+
+    // TODO(nv): Binary search.
+
+    auto it = _data.entries.begin();
+
+    auto chunk_ix = 0;
+
+    // If the first chunk is already greater than the requested NTP and offset,
+    // it means that the requested NTP and offset are not in the index.
+    if (unlikely(!within_bounds(ntp, offset))) {
+        return std::nullopt;
+    } else {
+        it++;
+    }
+
+    // Iterate until we find the first chunk that is greater than the requested
+    // NTP and offset and return the one before it.
+    while (it != _data.entries.end()) {
+        if (ntp < it->ntp || (it->ntp == ntp && offset < it->offset)) {
+            break;
+        }
+
+        chunk_ix++;
+        it++;
+    }
+
+    return _data.entries[chunk_ix].loc;
+}
+
+bool sparse_index::within_bounds(
```

Comment:
Can we replace sparse index with precise index that has every ntp?
So far, this looks like pessimisation. It's likely that the segment won't have millions of ntps. If it does we're in trouble even with the sparse index. We can set the limit on number of ntps that the segment can store. 
This will simplify the read path drastically, because you will be able to read the footer (the write intent may store the size of the footer for instance) and then you will be able to use regular record_batch_reader to scan the record batches. With sparse index you need to scan both record batches and some additional headers which indicate the beginning and the end of every ntp in the stream.